include/openssl/params.h

```diff
@@ -0,0 +1,119 @@
+# define OSSL_PARAM_uint(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(unsigned int), NULL)
+# define OSSL_PARAM_long(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_INTEGER, (addr), sizeof(long int), \
+                    NULL)
+# define OSSL_PARAM_ulong(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(unsigned long int), NULL)
+# define OSSL_PARAM_int32(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_INTEGER, (addr), sizeof(int32_t), NULL)
+# define OSSL_PARAM_uint32(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(uint32_t), NULL)
+# define OSSL_PARAM_int64(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_INTEGER, (addr), sizeof(int64_t), NULL)
+# define OSSL_PARAM_uint64(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(uint64_t), NULL)
+# define OSSL_PARAM_size_t(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), sizeof(size_t), \
+               NULL)
+# define OSSL_PARAM_double(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_REAL, (addr), sizeof(double), NULL)
+
+/* Basic parameter types including return sizes */
+# define OSSL_PARAM_SIZED_int(key, addr, sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_INTEGER, (addr), sizeof(int), &(sz))
+# define OSSL_PARAM_SIZED_uint(key, addr, sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(unsigned int), &(sz))
+# define OSSL_PARAM_SIZED_long(key, addr, sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_INTEGER, (addr), sizeof(long int), \
+                    &(sz))
+# define OSSL_PARAM_SIZED_ulong(key, addr, sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(unsigned long int), &(sz))
+# define OSSL_PARAM_SIZED_int32(key, addr, sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_INTEGER, (addr), sizeof(int32_t), &(sz))
+# define OSSL_PARAM_SIZED_uint32(key, addr, sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(uint32_t), &(sz))
+# define OSSL_PARAM_SIZED_int64(key, addr, sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_INTEGER, (addr), sizeof(int64_t), &(sz))
+# define OSSL_PARAM_SIZED_uint64(key, addr, sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(uint64_t), &(sz))
+# define OSSL_PARAM_SIZED_size_t(key, addr, sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), sizeof(size_t), \
+               &(sz))
+# define OSSL_PARAM_SIZED_double(key, addr, sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_REAL, (addr), sizeof(double), &(sz))
+
+# define OSSL_PARAM_SIZED_BN(key, addr, sz, r_sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), sz, \
+                    &(sz))
```

Comment:
That should be `&(r_sz)`

Reply:
Actually, I would rename `sz` to `r_sz` in all other `OSSL_PARAM_SIZED_TYPE` macros.  That makes `OSSL_PARAM_SIZED_BN` unusual by having an explicit buffer size, which is quite right, no?

Reply:
Unusual, yes.  I don't see how to avoid it.  We need to pass the size in to avoid potential overruns.

Using malloc is possible but has a few other issues which I'll need to discuss for the string and octet flavours (another PR though).

Reply:
What are you talking about?  I'm talking about the last macro parameter,  `r_sz`, which you are currently not using at all.

Reply:
I fixed that and used r_sz as the final argument to the macro.  I'm getting ahead of myself with the malloc talk.